### DISCUSSION TITLE
Graceful shutdown - Closing display window, using tart stop, or ctrl+c

### DISCUSSION THREAD

**Discussion by mbentley** [cirruslabs/tart#574] [Q&A]
I wasn't sure if this was something that was worth putting in an issue but it seems to me  that when I either close the display window (if one is active), use `tart stop...`, or use `ctrl+c`, the VM process is being killed instead of doing any sort of graceful shutdown.  I am not sure if this is anticipated but I have seen other solutions where it looks like they're using an agent running inside of the VM to initiate a clean, graceful shutdown.

Seems like if I am concerned about stateful data, this could lead to potential data issues.

**Comment by fkorotkov** [comment]
Right now Tart deliberately just request the VM to [stop](https://developer.apple.com/documentation/virtualization/vzvirtualmachine/3763076-stop). This is done purely for optimizing for scenario of creation one-off ephemeral VMs that you delete right away after running. There is a way to [request a stop](https://developer.apple.com/documentation/virtualization/vzvirtualmachine/3656700-requeststop) but in this case there will be just the window hanging:

<img width="1077" alt="Screenshot 2023-08-02 at 4 44 43 PM" src="https://github.com/cirruslabs/tart/assets/989066/d44fccc3-cb1d-411b-8ae7-5d66a9c5ac2e">

I'd recommend to incorporate `sudo -S shutdown -h now` similary to how Tart Packer Plugin does:

https://github.com/cirruslabs/packer-plugin-tart/blob/b25da725f429c5cddcb146f37e381596929efd2f/builder/tart/step_run.go#L105-L107

Alternatively, if you are running Sonoma host and guest via `tart run --resumable` then closing a window will actually suspend your VM. Check out [this video](https://twitter.com/cirrus_labs/status/1677308360385765382?s=20) on how it works.

**Comment by mbentley** [comment]
Thanks for that. I should have included that my use cases are all Linux based today, semi-stateful (trying to slowly move to stateless arm64 VMs for native Docker image builds).
